### PR TITLE
Get current user's nonprofit's object events

### PR DESCRIPTION
**Pull request by awadyas**
Closes https://github.com/CommitChange/tix/issues/3952
**NOTE: DO NOT discuss internal CommitChange information in your PR; this PR will be public.
Link back to the issue in the Tix repo when you need to do that.**

### CODE REVIEW

File: spec/requests/api_new/users_controller_request_spec.rb [hunk 1]
```diff
@@ -3,90 +3,103 @@
 
 describe ApiNew::UsersController, type: :request do
 
-
-	context 'for unlogged in user' do
-		it 'returns unauthorized when not logged in' do
-			get "/api_new/users/current"
-			expect(response).to have_http_status(:unauthorized)
+	context 'GET /api_new/users/current' do
```
Comment (wwahammy) [inline spec/requests/api_new/users_controller_request_spec.rb]: Please undo all of the changes that don't relate to the spec you're changing. It makes reviewing pretty hard

File: spec/requests/api_new/users_controller_request_spec.rb [hunk 2]
```diff
@@ -3,90 +3,103 @@
...
+		context 'for a nonprofit associate' do
+			let(:user) {create(:user_as_nonprofit_associate)}
 
+			before do
+				sign_in user
+				get "/api_new/users/#{user.id}/current_nonprofit_object_events"
+			end
 
-  
+			it "works" do
```
Comment (wwahammy) [inline spec/requests/api_new/users_controller_request_spec.rb]: Please be a bit more specific

File: spec/requests/api_new/users_controller_request_spec.rb [hunk 3]
```diff
@@ -3,90 +3,103 @@
...
-			)
-		}
-	end
-		
+		context 'for a nonprofit associate' do
+			let(:user) {create(:user_as_nonprofit_associate)}
 
+			before do
+				sign_in user
+				get "/api_new/users/#{user.id}/current_nonprofit_object_events"
```
Comment (wwahammy) [inline spec/requests/api_new/users_controller_request_spec.rb]: The route MUST not have the `user.id` in it. That's moving the original problem we were trying to fix.

Look at how `current` works in routes: we need this route to be: `/api_new/users/current_nonprofit_object_events`

File: spec/requests/api_new/users_controller_request_spec.rb [hunk 4]
```diff
@@ -3,90 +3,103 @@
...
+			end
 
-  
+			it "works" do
+				np_houid = user.roles.where(host_type: 'Nonprofit').first&.host&.houid
 
+				expect(response).to redirect_to "/api_new/nonprofits/#{np_houid}/object_events"
+				assert_redirected_to controller: 'api_new/object_events', action: 'index', nonprofit_id: np_houid
```
Comment (wwahammy) [inline spec/requests/api_new/users_controller_request_spec.rb]: These are redundant. `redirect_to` is the Rspec equivalent of `assert_redirected_to`. Please use the first line  but use options from the second.

File: spec/requests/api_new/users_controller_request_spec.rb [hunk 5]
```diff
@@ -3,90 +3,103 @@
...
-		it {
-			is_expected.to include_json(
-				object: 'user', 
-				is_super_admin: true,
-				roles: []
-			)
-		}
-	end
-		
+		context 'for a nonprofit associate' do
```
Comment (wwahammy) [inline spec/requests/api_new/users_controller_request_spec.rb]: Like mentioned before, you need to handle the situation where the user doesn't belong to any nonprofit.

File: app/controllers/api_new/users_controller.rb [hunk 6]
```diff
@@ -12,4 +12,9 @@ class ApiNew::UsersController < ApiNew::ApiController
 	def current
 		@user = current_user
 	end
+
+	# get /api_new/users/current_nonprofit/object_events
+	def current_nonprofit_object_events
+		redirect_to "/api_new/nonprofits/#{current_user.roles.where(host_type: 'Nonprofit').first&.host&.houid}/object_events"
```
Comment (wwahammy) [inline app/controllers/api_new/users_controller.rb]: Please use the Rails route helper method instead of the raw URL. Raw URLs are brittle because you can change your routes but not realize the URLs in the code are wrong.

File: app/controllers/api_new/users_controller.rb [hunk 7]
```diff
@@ -12,4 +12,9 @@ class ApiNew::UsersController < ApiNew::ApiController
 	def current
 		@user = current_user
 	end
+
+	# get /api_new/users/current_nonprofit/object_events
+	def current_nonprofit_object_events
```
Comment (wwahammy) [inline app/controllers/api_new/users_controller.rb:17]: As mentioned before the, this needs to handle the situation where the current user doesn't have a nonprofit

File: app/controllers/api_new/users_controller.rb (hunk 6 above)
Comment (wwahammy) [inline app/controllers/api_new/users_controller.rb]: You need to pass the parameters along as part of the redirect. The Rails route helper method makes that easier.

**Review by wwahammy — CHANGES_REQUESTED** [review]
Going in the right direction but there's a ways to go still.

File: app/controllers/api_new/users_controller.rb (hunk 7 above)
Comment (awadyas) [reply app/controllers/api_new/users_controller.rb:17]: What do u think should happen in that case? An error, or just redirect to a different page

File: spec/requests/api_new/users_controller_request_spec.rb [hunk 8]
```diff
@@ -83,10 +81,24 @@
...
-end
+			np_houid = user.roles.where(host_type: 'Nonprofit').first&.host&.houid
+			assert_redirected_to controller: 'api_new/object_events', action: 'index', nonprofit_id: np_houid
+		end
+
+		it "errors when user doesn't have a nonprofit" do
+			sign_in no_nonprofit_user
+			get current_nonprofit_object_events_api_new_users_path
 
+			expect(response).to have_http_status(500)
```
Comment (awadyas) [inline spec/requests/api_new/users_controller_request_spec.rb]: I'm guessing this isn't the best result for a user who doesn't have a nonprofit. What do u think should happen? @wwahammy

Comment (wwahammy) [reply spec/requests/api_new/users_controller_request_spec.rb]: I think a 404 would be appropriate. The check should be:

`expect(response).to have_http_status(:not_found)` I think

File: spec/requests/api_new/users_controller_request_spec.rb [hunk 9]
```diff
@@ -36,7 +34,7 @@
 					}
 			])
 		}
-
```
Comment (wwahammy) [inline spec/requests/api_new/users_controller_request_spec.rb]: Please remove any unrelated formatting changes.

File: app/controllers/api_new/users_controller.rb [hunk 10]
```diff
@@ -12,4 +12,14 @@ class ApiNew::UsersController < ApiNew::ApiController
...
+		np_houid = current_user.roles.where(host_type: 'Nonprofit').first&.host&.houid
+		if np_houid
+			redirect_to api_new_nonprofit_object_events_path(np_houid)
```
Comment (wwahammy) [inline app/controllers/api_new/users_controller.rb]: Any get query params need to be passed. The route helper has the ability accept those params as arguments if you pass them to it.

Additionally, you need to test that provided params are correctly passed.

**Review by wwahammy — CHANGES_REQUESTED** [review]

**Review by wwahammy — APPROVED** [review]
Looks good!
